### PR TITLE
[optimize](count) optimize key exact query without reading data

### PR DESCRIPTION
**Pull request by zzzxl1993**
## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 1]
```diff
@@ -2570,5 +2573,64 @@ void SegmentIterator::_calculate_pred_in_remaining_conjunct_root(
...
+            std::string res;
+            if (filed_type == FieldType::OLAP_FIELD_TYPE_DATETIMEV2) {
+                auto max = type_limit<PrimitiveTypeTraits<TYPE_DATETIMEV2>::CppType>::max();
+                res.resize(30);
+                max.to_string(res.data());
+            } else if (filed_type == FieldType::OLAP_FIELD_TYPE_INT) {
+                auto max = type_limit<PrimitiveTypeTraits<TYPE_INT>::CppType>::max();
+                res = boost::lexical_cast<std::string>(max);
+            }
+            return res;
```
Comment (airborne12) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:2602]: how about other type of field?

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 2]
```diff
@@ -2570,5 +2573,64 @@ void SegmentIterator::_calculate_pred_in_remaining_conjunct_root(
     }
 }
 
+bool SegmentIterator::_need_read_pk_data(ColumnId cid, vectorized::MutableColumnPtr& column,
+                                         size_t nrows_read) {
+    if (_opts.tablet_schema->keys_type() != KeysType::DUP_KEYS) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:2578]: Is it OK for MOW?

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 3]
```diff
@@ -2570,5 +2573,64 @@ void SegmentIterator::_calculate_pred_in_remaining_conjunct_root(
...
+        return false;
+    }
+
+    if (_opts.push_down_agg_type_opt != TPushAggOp::COUNT_ON_INDEX) {
+        return false;
+    }
+
+    if (!_opts.tablet_schema->column(cid).is_key()) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:2586]: If the table key is k1,k2,k3, the column for cid is k2, is it right?

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 4]
```diff
@@ -2570,5 +2573,64 @@ void SegmentIterator::_calculate_pred_in_remaining_conjunct_root(
...
+                    const auto& pk_value = upper_tuple.get_value(cid);
+                    if (pk_max != pk_value) {
+                        return false;
+                    }
+                }
+            }
+        }
+    }
+
+    if (column->is_nullable()) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:2624]: Do you mean use defaults for not read data?

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp [hunk 5]
```diff
@@ -2570,5 +2573,64 @@ void SegmentIterator::_calculate_pred_in_remaining_conjunct_root(
...
+        // 2. upper is not null and is the maximum value
+        for (auto& key_range : _opts.key_ranges) {
+            auto lower_tuple = key_range.lower_key->to_tuple();
+            auto upper_tuple = key_range.upper_key->to_tuple();
+            for (size_t cid = 0; cid < lower_tuple.size(); cid++) {
+                if (lower_tuple.is_null(cid) && !upper_tuple.is_null(cid)) {
+                    const auto* field = key_range.upper_key->column_schema(cid);
+                    auto pk_max = get_max_datetime_string(field->type());
+                    const auto& pk_value = upper_tuple.get_value(cid);
+                    if (pk_max != pk_value) {
```
Comment (xiaokang) [inline be/src/olap/rowset/segment_v2/segment_iterator.cpp:2616]: It is different from the comment.

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

**Comment by zzzxl1993** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_iterator.cpp (hunk 3 above)
Comment (xiaokang) [reply be/src/olap/rowset/segment_v2/segment_iterator.cpp:2591]: should be prefix of table keys

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by airborne12 — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**